hydra-node/src/Hydra/Network/Heartbeat.hs

```diff
@@ -1,63 +1,173 @@
+  (MonadSTM m, MonadMonotonicTime m) =>
   TVar m HeartbeatState ->
-  Network m (HydraMessage tx)
-checkMessages Network{broadcast} heartbeatState =
+  NetworkCallback (HydraMessage msg) m ->
+  NetworkCallback (Heartbeat (HydraMessage msg)) m
+updateStateFromIncomingMessages heartbeatState callback = \case
+  Message msg -> notifyAlive (getParty msg) >> callback msg
+  Ping party -> notifyAlive party
+ where
+  notifyAlive party = do
+    now <- getMonotonicTime
+    aliveSet <- alive <$> readTVarIO heartbeatState
+    unless (party `Map.member` aliveSet) $ callback (Connected party)
+    atomically $
+      modifyTVar' heartbeatState $ \s ->
+        s
+          { alive = Map.insert party now (alive s)
+          , suspected = party `Set.delete` suspected s
+          }
+
+updateStateFromOutgoingMessages ::
+  (MonadSTM m, MonadMonotonicTime m) =>
+  TVar m HeartbeatState ->
+  Network m (Heartbeat (HydraMessage msg)) ->
+  Network m (HydraMessage msg)
+updateStateFromOutgoingMessages heartbeatState Network{broadcast} =
   Network $ \msg -> do
-    case msg of
-      Ping _ -> pure ()
-      _ -> atomically (writeTVar heartbeatState StopHeartbeat)
-    broadcast msg
+    now <- getMonotonicTime
+    updateLastSent heartbeatState now
+    broadcast (Message msg)
+
+updateLastSent :: MonadSTM m => TVar m HeartbeatState -> Time -> m ()
+updateLastSent heartbeatState now = atomically (modifyTVar' heartbeatState $ \s -> s{lastSent = Just now})
 
-sendHeartbeatFor ::
+checkHeartbeatState ::
   ( MonadDelay m
   , MonadSTM m
+  , MonadMonotonicTime m
   ) =>
-  Host ->
+  Party ->
   TVar m HeartbeatState ->
-  Network m (HydraMessage tx) ->
+  Network m (Heartbeat (HydraMessage msg)) ->
   m ()
-sendHeartbeatFor localhost heartbeatState Network{broadcast} =
+checkHeartbeatState localhost heartbeatState Network{broadcast} =
   forever $ do
-    threadDelay 0.5
+    threadDelay heartbeatDelay
     st <- readTVarIO heartbeatState
-    when (st == SendHeartbeat) $ broadcast (Ping localhost)
+    now <- getMonotonicTime
+    when (shouldSendHeartbeat now st) $ do
+      updateLastSent heartbeatState now
+      broadcast (Ping localhost)
+
+shouldSendHeartbeat :: Time -> HeartbeatState -> Bool
+shouldSendHeartbeat now HeartbeatState{lastSent} =
+  maybe True (checkTimeout id now) lastSent
+
+checkRemoteParties ::
+  ( MonadDelay m
+  , MonadSTM m
+  , MonadMonotonicTime m
+  ) =>
+  TVar m HeartbeatState ->
+  NetworkCallback (HydraMessage msg) m ->
+  m ()
+checkRemoteParties heartbeatState callback =
+  forever $ do
+    threadDelay (heartbeatDelay * 2)
+    now <- getMonotonicTime
+    updateSuspected heartbeatState now
+      >>= mapM_ (callback . Disconnected)
+
+updateSuspected :: MonadSTM m => TVar m HeartbeatState -> Time -> m (Set Party)
+updateSuspected heartbeatState now =
+  atomically $ do
+    aliveParties <- alive <$> readTVar heartbeatState
+    let timedOutParties = Map.filter (checkTimeout (2 *) now) aliveParties
```

Comment:
Analogous to `shouldSendHeartbeat`, a function with something like

```
checkTimedOut now lastSeen = now > lastSeen + 2 * livenesDelay
```

Would be easier to read and understand IMO (instead of passing a modifier to a re-used `checkTimeout` function.